pkg/controller/persistentvolume/types.go

```diff
@@ -75,23 +74,38 @@ type matchPredicate func(compareThis, toThis *api.PersistentVolume) bool
 
 // Find returns the nearest PV from the ordered list or nil if a match is not found
 func (pvIndex *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matchPredicate matchPredicate) (*api.PersistentVolume, error) {
-	volumes, err := pvIndex.ListByAccessModes(pv.Spec.AccessModes)
-	if err != nil {
-		return nil, err
-	}
+	// the 'pv' argument is a synthetic PV with capacity and accessmodes set according to the user's PersistentVolumeClaim.
+	// the synthetic pv arg is, therefore, a request for a storage resource.
+	//
+	// PVs are indexed by their access modes to allow easier searching.  Each index is the string representation of a set of access modes.
+	// There is a finite number of possible sets and PVs will only be indexed in one of them (whichever index matches the PV's modes).
+	//
+	// A request for resources will always specify its desired access modes.  Any matching PV must have at least that number
+	// of access modes, but it can have more.  For example, a user asks for ReadWriteOnce but a GCEPD is available, which is ReadWriteOnce+ReadOnlyMany.
+	//
+	// Searches are performed against a set of access modes, so we can attempt not only the exact matching modes but also
+	// potential matches (the GCEPD example above).
+	allPossibleModes := pvIndex.allPossibleMatchingAccessModes(pv.Spec.AccessModes)
```

Comment:
@saad-ali Is there a way to make this current implementation more clear?  

What you are proposing is a rewrite of the PV index.  That is, to me, riskier and more time consuming than this issue warrants.

I think, too, there is an equivalence here.

1.  Figure out all the possible combinations of access modes a PV can have (or are requested) and index the PV in each.  This allows us to query the index once.

2.  Figure out all possible combinations that satisfy a request for a PV and query the index for each.  Keep the PV in a single index.

Either way, It seems like one piece has to go from 1 to Many.  I choose the second option because it was simpler than trying to rewrite the index, which is riskier, IMHO.

Thoughts?